external/src/main/java/com/redhat/cloud/policies/engine/actions/plugins/NotificationActionPluginListener.java

```diff
@@ -44,72 +47,107 @@
     public static final String APP_NAME = "policies";
     public static final String EVENT_TYPE_NAME = "policy-triggered";
 
+    private final Logger log = Logger.getLogger(this.getClass().getName());
+    private final ConcurrentSkipListMap<String, PoliciesAction> notifyBuffer = new ConcurrentSkipListMap<>();
 
     @Inject
     @Channel("webhook")
     @OnOverflow(OnOverflow.Strategy.UNBOUNDED_BUFFER)
     Emitter<String> channel;
 
     @Inject
-    @Metric(absolute = true, name = "engine.actions.webhook.processed")
+    @Metric(absolute = true, name = "engine.actions.notifications.processed")
     Counter messagesCount;
 
+    @Inject
+    @Metric(absolute = true, name = "engine.actions.notifications.aggregated")
+    Counter messagesAggregated;
+
+    @Inject
+    ObjectMapper objectMapper;
 
     @Override
     public void process(ActionMessage actionMessage) throws Exception {
         messagesCount.inc();
-        Action notificationAction = new Action();
-        notificationAction.setEventType(EVENT_TYPE_NAME);
-        notificationAction.setApplication(APP_NAME);
-        notificationAction.setBundle(BUNDLE_NAME);
-        notificationAction.setTimestamp(
-                LocalDateTime.ofInstant(Instant.ofEpochMilli(actionMessage.getAction().getCtime()), ZoneId.systemDefault())
+        PoliciesAction policiesAction = new PoliciesAction();
+        policiesAction.setAccountId(actionMessage.getAction().getTenantId());
+        policiesAction.setTimestamp(
+            LocalDateTime.ofInstant(Instant.ofEpochMilli(actionMessage.getAction().getCtime()), ZoneId.systemDefault())
         );
 
-        PoliciesPayloadBuilder payloadBuilder = new PoliciesPayloadBuilder();
+        PoliciesAction.Context context = policiesAction.getContext();
+        Set<PoliciesAction.Event> events = policiesAction.getEvents();
 
         for (Map.Entry<String, String> tagEntry : actionMessage.getAction().getEvent().getTags().entries()) {
             String value = tagEntry.getValue();
             if (value == null) {
                 // Same behavior as previously with JsonObjectNoNullSerializer with old hooks
                 value = "";
             }
-            payloadBuilder.addTag(tagEntry.getKey(), value);
-        }
 
-        notificationAction.setAccountId(actionMessage.getAction().getTenantId());
+            if (!context.getTags().containsKey(tagEntry.getKey())) {
+                context.getTags().put(tagEntry.getKey(), new HashSet<>());
+            }
```

Comment:
```suggestion
            context.getTags().putIfAbsent(tagEntry.getKey(), new HashSet<>());
```

Reply:
I'll use `computeIfAbsent`. Wasn't aware of these methods. Thanks!